### DISCUSSION TITLE
Does SF2 require a specific ordering for instrument generators?

### DISCUSSION THREAD

**Discussion by IAGOnGithub** [FluidSynth/fluidsynth#1359] [General]
Hi all,

I'm trying to refactor a SF2 generation tool named GBA Mus Ripper, and am brushing up against an issue where the generated SF2 files do not work correctly with Fluidsynth (x86_64 GNU/Linux Arch Linux); specifically, Fluidsynth ends up discarding some "invalid instrument generators". They can be fixed automatically (and "magically") by importing them into the Soundfont editor Polyphone. However, comparing the IGEN chunks of the "fixed" and "broken" SF2 files I found that the only differences between them are in how the generators are ordered. I can't understand why this would have an impact, so long as the sampleId generators are the last ones in the zone.

Here are all of the files needed to reproduce: [reproducing.zip](https://github.com/user-attachments/files/16386279/reproducing.zip)

This may be a bug. I am able to open a bug report if need be, but wanted to make sure it actually was one first.

**Comment by IAGOnGithub** [comment]
From the SoundFont 2.0 specification document, section 8.1.2, in regards to the keyRange generator (numbered 43):
> This is the minimum and maximum MIDI key number values for which this preset zone or instrument zone is active. The LS byte indicates the highest and the MS byte the lowest valid key. The keyRange enumerator is optional, but when it does appear, **it must be the first generator in the zone generator list.**

It's a GBA Mus Ripper bug. I will fix it.

**Reply by IAGOnGithub** [reply]
Update: I fixed it.

**Reply by derselbst** [reply]
Thanks for taking care of this!
